Include/cpython/sysmodule.h

```diff
@@ -19,7 +19,7 @@ PyAPI_FUNC(int) PyUnstable_PerfMapState_Init(void);
 PyAPI_FUNC(int) PyUnstable_WritePerfMapEntry(
     const void *code_addr,
     unsigned int code_size,
-    const char *entry_name);
+    PyCodeObject *entry_name);
```

Comment:
@gsallam  I needed to change the API of some of these functions because they were operating on the main "perf" trampoline and they were not compatible with multiple possible trampolines. Please take a look and tell me if you see anything we should re-consider here.

Reply:
@pablogsal The purpose of this function is that it should be entirely "lower level" than the perf-trampoline feature. There can inherently be only one perfmap file per process (by the definition of the perfmap file location), and this function should always write an entry to that file (in a thread-safe way). And for such a low level function, it doesn't make sense to tie it to code objects, either; the function should just accept the raw data needed to write an entry to the perf map file.

So I don't really understand why either the signature or the implementation of this function should change in this PR. This function should not be coupled to anything related to perf trampoline callbacks, etc; it should always simply write one entry to `/tmp/perf-<pid>.map`.

Reply:
Yeah I understand that but the key is if you want to maintain the `PyUnstable_WritePerfMapEntry` tied to the "perf" version of the trampolines or to *any* version of the trampolines. For instance, with the `perfjit` version that works with DWARF you may want exactly the same behaviour but with this version you don't write to `/tmp/perf-<pid>.map` but to a different file with a different format.

The other thing to consider is that these functions (like `PyUnstable_PerfMapState_Init`) are being called from the interpreter as if they are general for any trampoline but there is some entanglement where they assume only one trampoline version exist. We can leave the `PyUnstable_PerfMap...` function as low level only for the perf trampoline and make generic versions that we use internally or make the `PyUnstable_PerfMap...` work with *any* trampoline.

I do apologise if this is a bit confusing but unfortunately there is some tanglement currently and is a bit challenging to explain it if you don't look at the code. I can explain in more detail if this is not enough :)

Reply:
I don't think `PyUnstable_WritePerfMapEntry` should be tied to _any_ version of the trampoline feature. It can be called by the `perf` version of the trampoline feature. It should continue to have exactly one use, which it was added for, which is to write (in a thread-safe way, potentially by multiple unrelated callers) to a file in `/tmp/perf-<pid>.map`, for use by `perf`.

If we want to introduce support for writing to some other similar file with a different format, that should get its own new and separate low-level thread-safe writer function, which can in turn be used by the new variant of the trampoline feature.

Reply:
Ok, I acquired slightly more of an opinion :)

I think the term "PerfMap" should continue to refer only to "perf map" files, i.e. the files in `tmp/perf-<pid>.map`, and all of the API in `sysmodule.c` devoted to working with those files (that is, `PyUnstable_PerfMapState_Init`, `PyUnstable_WritePerfMapEntry`, `PyUnstable_CopyPerfMapFile`, and `PyUnstable_PerfMapState_Fini`) should be left alone. This API is not part of the trampoline feature; it is a lower level API which can be used by the trampolines, but is also intended for use by other unrelated callers (e.g. the Cinder JIT.) It is necessary for this API to exist so that multiple independent clients can all independently and safely write to the perf-map file, possibly from multiple threads. These other external callers should be able to use the perf map API and not be tied in any way to the configuration (or even the use) of the trampoline feature.

If we now want to also add support for writing "jit dump" files, we should add new API (using the term "JitDump" rather than "PerfMap") for that separate purpose, which can also be used by external callers who want to safely write their own entries to the jit dump file. That API should also be low level and not tied in any way to the trampoline feature. But of course the trampoline feature can use it to add jit-dump support.

Reply:
Ok, I think I agree with your vision here. I will try to modify this patch to see if I can bend it to this design and will report any challenges I find (if any). Thanks a lot for your input @carljm!